common/src/main/java/org/conscrypt/HpkeSuite.java

```diff
@@ -0,0 +1,330 @@
+   *      <ul>KDF  : 0x0001: HKDF-SHA256</ul>
+   *      <ul>AEAD : 0x0003: ChaCha20Poly1305</ul>
+   * </li>
+   *
+   * @see <a
+   * href="https://www.rfc-editor.org/rfc/rfc9180
+   * .html#name-key-encapsulation-mechanism">KEMs</a>
+   * @see <a
+   * href="https://www.rfc-editor.org/rfc/rfc9180
+   * .html#name-key-derivation-functions-kd">KDFs</a>
+   * @see <a
+   * href="https://www.rfc-editor.org/rfc/rfc9180
+   * .html#name-authenticated-encryption-wi">AEAD</a>
+   */
+  public static final HpkeSuite DHKEM_X25519_HKDF_SHA256_HKDF_SHA256_CHACHA20POLY1305 =
+      new HpkeSuite(KEM.DHKEM_X25519_HKDF_SHA256, KDF.HKDF_SHA256, AEAD.CHACHA20POLY1305);
+
+  private final KEM mKem;
+  private final KDF mKdf;
+  private final AEAD mAead;
+
+  private HpkeSuite(KEM Kem, KDF Kdf, AEAD Aead) {
+    mKem = Kem;
+    mKdf = Kdf;
+    mAead = Aead;
+  }
+
+  /**
+   * KEM configured while creating an instance of {@link HpkeSuite}
+   *
+   * @return kem
+   */
+  KEM getKem() {
+    return mKem;
+  }
+
+  /**
+   * KDF configured while creating an instance of {@link HpkeSuite}
+   *
+   * @return kdf
+   */
+  KDF getKdf() {
+    return mKdf;
+  }
+
+  /**
+   * AEAD configured while creating an instance of {@link HpkeSuite}
+   *
+   * @return aead
+   */
+  AEAD getAead() {
+    return mAead;
+  }
+
+  enum KEM {
+    DHKEM_X25519_HKDF_SHA256(
+        /* id= */ 32, /* encLength= */ 32, /* pkLength= */ 32, /* skLength= */ 32);
+
+    private final int id;
+    private final int encLength;
+    private final int pkLength;
+    private final int skLength;
+
+    KEM(int id, int encLength, int pkLength, int skLength) {
+      this.id = id;
+      this.encLength = encLength;
+      this.pkLength = pkLength;
+      this.skLength = skLength;
+    }
+
+    /**
+     * KEM id in its decimal representation
+     *
+     * @return kem id
+     * @see <a
+     *         href="https://www.rfc-editor.org/rfc/rfc9180.html#name-key-encapsulation-mechanism">KEM
+     *         ids</a>
+     */
+    int getId() {
+      return id;
+    }
+
+    /**
+     * The length in bytes of an encapsulated key produced by this KEM.
+     *
+     * @return encapsulated key size in bytes
+     */
+    int getEncLength() {
+      return encLength;
+    }
+
+    /**
+     * The length in bytes of an encoded public key for this KEM.
+     *
+     * @return public key size in bytes
+     */
+    int getPkLength() {
+      return pkLength;
+    }
+
+    /**
+     * The length in bytes of an encoded private key for this KEM.
+     *
+     * @return private key size in bytes
+     */
+    int getSkLength() {
+      return skLength;
+    }
+
+    /**
+     * Validates the enc size in bytes matches the {@link KEM} spec.
+     *
+     * @param enc encapsulated key produced by the kem
+     * @see <a
+     *         href="https://www.rfc-editor.org/rfc/rfc9180.html#name-key-encapsulation-mechanism">expected
+     *         enc size</a>
+     */
+    void validateEnc(byte[] enc) {
```

Comment:
The name of this function is misleading. It reads like you're validating `enc`, but in reality you're only validating the length of `enc`. There are other things that can cause an `enc` value to be rejected. (For P-256, the curve may not be valid. For X25519, it could be a low-order point.)

BoringSSL already performs these checks. Any reason not to do that internally?

Reply:
I have updated the name to perform a quick validation only on the length. I was thinking of having an initial validation to avoid calling the JNI functions if the length doesn't match.

Reply:
Mm. I probably would have just left it to the C/C++ bits, so it's less code that we need to check for correctness, but don't feel strongly. I guess you intend this as a perf optimization?